src/main/java/picard/illumina/IlluminaBasecallsToFastq.java

```diff
@@ -486,17 +529,48 @@ public void write(final ClusterData rec) {
                         throw new IllegalStateException("Read type other than T/B/M encountered.");
                 }
 
-                writeSingle(out, name, read);
+                writeSingle(out, name, read, templateIndex);
             }
         }
 
         /**
          * Writes out a single read to a single FASTQ's output stream.
          */
-        private void writeSingle(final OutputStream out, final String name, final ReadData read) {
+        private void writeSingle(final OutputStream out, final String name, final ReadData read, int templateIndex) {
             try {
-                final byte[] bases = read.getBases();
-                final byte[] quals = read.getQualities();
+                byte[] bases = read.getBases();
+                byte[] quals = read.getQualities();
+
+                if (trimmingQuality != null) {
+                    int index = TrimmingUtil.findQualityTrimPoint(quals, trimmingQuality);
+                    // Don't write zero length reads if the entire read is trimmed
+                    if (index == 0) {
+                        return;
+                    }
+                    quals = Arrays.copyOfRange(quals, 0, index);
+                    bases = Arrays.copyOfRange(bases, 0, index);
+                }
```

Comment:
So I don't think we can do the `if (index == 0)` test here.  If we have PE reads, or even say reads with extracted barcodes or UMIs being written to separate fastqs, then we could skip e.g. R1 but include R2 and cause record/line numbers to go out of sync.  

Also is this applied to all reads regardless of type?  I think we should not be adapter or quality trimming anything but template reads.

I think I would suggest moving the trimming-related code so that it's called a) only on template reads and b) directly from write(ClusterData).  That way we can e.g. skip over clusters where all reads are trimmed to 0bp, and otherwise do something more sensible.  Perhaps there's a `MIN_TRIMMED_LENGTH` or similar parameter, that controls this?  I.e. if that's set to 20bp and quality trimming says remove all of it (but not it's mate), then we only trim to 20bp?

Reply:
I think the adapter marking would also be better too since you should in theory find the adapter at the same base in each read and that boosts confidence.